### DISCUSSION TITLE
How to set_adapter in the forward function?

### DISCUSSION THREAD

**Discussion by Weibin44** [huggingface/peft#1924] [Q&A]
I know the basic usage of peft, but I want to implement a more detail thing:
Outside of `Trainer `function, I have built a model and added two adapter, but how to switch these adapters in the forward function? I want to use these two adapters parallel. For example, I have two inputs, 1 and 2, for `input_1`, I use `adapter_1`, for `input_2 `I use `adapter_2`, the `adapter_2 `doesn't participate in the calculation of `input_1`, and so does `adapter_1`. After I got these two outputs, I use them as the input for another model. 

I tried to call `self.set_adapter("adapter_name")` in the forward function but got `ValueError: No adapter loaded. Please load an adapter first.` I think it might because the object type is different in the forward?

Thanks for your help in advance! every valuable suggestions are greatly appreciated!

**Comment by Weibin44** [comment]
When I tried to add adapters in the model __init__, I got this error:
`During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 619, in __getattr__
    return super().__getattr__(name)  # defer to nn.Module's logic
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'PeftModelForCausalLM' object has no attribute 'base_model'`

**Comment by BenjaminBossan** [comment]
This is very hard to reply to without seeing some actual code. Could you please show the essential pieces of code that you have, how you call it, and the full error message?

**Reply by Weibin44** [reply]
Hi BenjaminBossan, thanks for your prompt reply!
The problem seems solved now. I am testing whether it works correctly now. 
By the way, can I confirm how `model.set_adapter("B")` works? does this function mean only the `adapter B` works on forward and backward, and any other adapters are removed temporarily? (no influence on both forward and backward) Therefore If I want to apply different adapters for different kinds of data during training, I can:
```
class MyModel(LlamaForCausalLM):
    def __init__(self, config, **kwargs):
        super(LlamaForCausalLM, self).__init__(config)
        self.model = LlamaModel.from_pretrained(config)
        lora_config = LoraConfig(
            r=kwargs["lora_r"],
            lora_alpha=kwargs["lora_alpha"],
            target_modules=find_all_linear_names(self.model),
            lora_dropout=kwargs["lora_dropout"],
            bias=kwargs["lora_bias"],
            task_type="FEATURE_EXTRACTION",
        )
        self.model = get_peft_model(self.model, lora_config, adapter_name="A")
        self.model.add_adapter(adapter_name="B", peft_config=lora_config)

    def forward(self, data_A, data_B):
            self.model.set_adapter("A")
            self.model(data_A) # only adapter A participate in forward and backward
            self.model.set_adapter("B")
            self.model(data_B) # only adapter B participate in forward and backward. The gradients are independent between A and B.
```
I am not sure if I understand it correctly. Thanks for your help in advance!

**Reply by BenjaminBossan** [reply]
In general, the code that you showed looks valid. One thing to note, however, is that `Trainer` only passes the parameters to the optimizer that have `requires_grad=True`. Therefore, when the model is initialized, only adapter A will have its parameters passed to the optimizer, which means that B is never updated. To fix this, try adding this line after `self.model.add_adapter(adapter_name="B", peft_config=lora_config)`:

```python
self.model.base_model.set_adapter(["A", "B"])
```

The idea is to temporarily activate both adapters, so that the `Trainer` adds both their parameters to the optimizer. But please verify if that works correctly by checking the `trainer.optimizer` or by training for one step and then checking that both adapters were updated.

That said, I have never tried training two adapters alternatively in the same run, so there could be other issues I'm currently not seeing. It's always good to train only a little bit and then verifying that both adapters have been updated as expected.

Regarding the error you reported, could you please post the full error message? Are you using distributed training like DeepSpeed or FSDP? If yes, could you try installing the latest PEFT version from the `main` branch?

**Reply by Weibin44** [reply]
Thanks for your informative reply! Let me test it later.

Yes, I am using DeepSpeed, and my peft is 0.11.1, it should be the latest version? My previous error is here:
```
....
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 619, in __getattr__
    return super().__getattr__(name)  # defer to nn.Module's logic
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'PeftModelForFeatureExtraction' object has no attribute 'base_model'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 619, in __getattr__
    return super().__getattr__(name)  # defer to nn.Module's logic
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'PeftModelForFeatureExtraction' object has no attribute 'base_model'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 619, in __getattr__
    return super().__getattr__(name)  # defer to nn.Module's logic
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'PeftModelForFeatureExtraction' object has no attribute 'base_model'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 619, in __getattr__
    return super().__getattr__(name)  # defer to nn.Module's logic
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'PeftModelForFeatureExtraction' object has no attribute 'base_model'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/weibin/meme/LLaVA/llava/train/train_xformers.py", line 15, in <module>
    train()
  File "/home/weibin/meme/LLaVA/llava/train/train.py", line 920, in train
    model = LlavaLlamaForCausalLM.from_pretrained(
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/transformers/modeling_utils.py", line 3596, in from_pretrained
    model = cls(config, *model_args, **model_kwargs)
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 459, in wrapper
    f(module, *args, **kwargs)
  File "/home/weibin/meme/LLaVA/llava/model/language_model/llava_llama.py", line 91, in __init__
    self.model = get_peft_model(self.model, lora_config, adapter_name="Decoder")
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/mapping.py", line 149, in get_peft_model
    return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](model, peft_config, adapter_name=adapter_name)
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 454, in wrapper
    if not hasattr(module, "_ds_child_entered"):
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 621, in __getattr__
    return getattr(self.base_model, name)
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 621, in __getattr__
    return getattr(self.base_model, name)
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 621, in __getattr__
    return getattr(self.base_model, name)
  [Previous line repeated 985 more times]
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/peft/peft_model.py", line 619, in __getattr__
    return super().__getattr__(name)  # defer to nn.Module's logic
  File "/home/weibin/anaconda3/envs/llava_official/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
RecursionError: maximum recursion depth exceeded while calling a Python object
```
This error happens when I called `model = LlavaLlamaForCausalLM.from_pretrained`, the `LlavaLlamaForCausalLM `is here, this is based on the source code from [the official](https://github.com/haotian-liu/LLaVA):
```
class LlavaLlamaForCausalLM(LlamaForCausalLM, LlavaMetaForCausalLM):
    config_class = LlavaConfig
    # def __init__(self, configname, cache_dir, attn_implementation, torch_dtype,**kwargs):
    def __init__(self, config, **kwargs):
        super(LlamaForCausalLM, self).__init__(config)
        self.model = LlavaLlamaModel(config)
        lora_config = LoraConfig(
            r=kwargs["lora_r"],
            lora_alpha=kwargs["lora_alpha"],
            target_modules=find_all_linear_names(self.model),
            lora_dropout=kwargs["lora_dropout"],
            bias=kwargs["lora_bias"],
            task_type="FEATURE_EXTRACTION",
        )
        self.model = get_peft_model(self.model, lora_config, adapter_name="A")
```

**Reply by Weibin44** [reply]
After I met this error, I changed the code to the structure in the first paragraph, where I simply initial `LlavaLlamaForCausalLM` outside, and load `self.model.from_pretrained`. Then the error seems removed (but I don't know why). I am new to this, it seems there are many operating mechanisms hidden in it, and the official simple tutorial is not enough for me. I hope there will be more suitable materials for me to know and become familiar with it. Thanks for your help again :)

**Reply by Weibin44** [reply]
I checked it by using safe_get_full_fp32_param from deepspeed, both adapters were updated. However, once I want to print the weights of `self.model`, it raised `AttributeError: 'NoneType' object has no attribute 'shape'`:
```
for n, lp in self.model.named_parameters():
                if "0.self_attn.k_proj" in n:
                    p = safe_get_full_fp32_param(lp)
                    print(p, p.shape, n)
```
This error also raises if I didn't use `self.model.base_model.set_adapter(["A", "B"])` but want to print the weights of `adapter B` in forward.

**Reply by BenjaminBossan** [reply]
Regarding the recursion error, it should be solved when you use the latest version of PEFT installed from main. E.g. run this in your virtual env:

`python -m pip install git+https://github.com/huggingface/peft`

> However, once I want to print the weights of `self.model`, it raised `AttributeError: 'NoneType' object has no attribute 'shape'`

When you're still running DeepSpeed, this error is most likely caused by the weights being sharded to another machine. The machine that raises the error only has a placeholder `None`. To access it on that machine, you need to use the `deepspeed.zero.GatheredParameters` context from DeepSpeed.

**Reply by Weibin44** [reply]
Thanks for your informative reply!
